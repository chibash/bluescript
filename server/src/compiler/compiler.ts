import { GlobalVariableNameTable } from "../transpiler/code-generator/variables";
import { transpile } from "../transpiler/code-generator/code-generator";
import * as fs from "fs";
import * as path from "path";
import {Buffer} from "node:buffer";
import { spawn } from "child_process";
import generateLinkerScript from "./linker-script";
import generateMakefile from "./makefile";
import { ElfReader } from "./elf-reader";


export type PackageConfig = {
    name: string,
    espIdfComponents: string[],
    dependencies: string[],
    dirs: {
        root: string,
        dist: string,
        build: string,
        packages: string,
    }
}

export type CompilerConfig = {
    dirs: {
        runtime: string,
        compilerToolchain: string,
        std: string
    }
};

export type MemoryLayout = {
    iram:{address:number, size:number},
    dram:{address:number, size:number},
    iflash:{address:number, size:number},
    dflash:{address:number, size:number},
}


type MemoryRegion = {name: string, address: number, size: number}

export class ShadowMemory {
    public iram: MemoryRegion;
    public dram: MemoryRegion;
    public iflash: MemoryRegion;
    public dflash: MemoryRegion;

    constructor(memoryLayout: MemoryLayout) {
        this.iram = {name: '.iram', ...memoryLayout.iram};
        this.dram = {name: '.dram', ...memoryLayout.dram};
        this.iflash = {name: '.iflash', ...memoryLayout.iflash};
        this.dflash = {name: '.dflash', ...memoryLayout.dflash};
    }
}

export type ExecutableBinary = {
    iram?: {address: number, data: Buffer},
    dram?: {address: number, data: Buffer},
    iflash?: {address: number, data: Buffer},
    dflash?: {address: number, data: Buffer},
    entryPoints: {id: number, address: number}[]
}

const MODULE_PATH = (pkg: PackageConfig, relativePath: string) => path.normalize(path.join(pkg.dirs.root, `${relativePath}.bs`));
const MODULE_C_PATH = (dir: string, moduleName: string) => path.join(dir, `bs_${moduleName}.c`);
const ARCHIVE_PATH = (pkg: PackageConfig) => path.join(pkg.dirs.build, `lib${pkg.name}.a`);
const MAKEFILE_PATH = (pkg: PackageConfig) => path.join(pkg.dirs.dist, 'Makefile');
const LINKER_SCRIPT = (pkg: PackageConfig) => path.join(pkg.dirs.build, "linkerscript.ld");
const LINKED_ELF_PATH = (pkg: PackageConfig) => path.join(pkg.dirs.build, `${pkg.name}.elf`);
const STD_MODULE_PATH = (compilerConfig: CompilerConfig) => path.join(compilerConfig.dirs.std, 'index.bs');
const LD_PATH = (compilerConfig: CompilerConfig) => path.join(compilerConfig.dirs.compilerToolchain, 'xtensa-esp32-elf-ld');
const RUNTIME_ELF_PATH = (compilerConfig: CompilerConfig) => path.join(compilerConfig.dirs.runtime, 'ports/esp32/build/bluescript.elf');
const C_PROLOG = (compilerConfig: CompilerConfig) => `
#include <stdint.h>
#include "${path.join(compilerConfig.dirs.runtime, '/core/include/c-runtime.h')}"

`

export class Compiler {
    private config: CompilerConfig;
    private espidfComponents: ESPIDFComponents;
    private memory: ShadowMemory;
    private packageReader: (name: string) => PackageConfig;
    private logger: {
            info: (message: string) => void,
            error: (message: string) => void
    };

    constructor(
        memoryLayout: MemoryLayout, 
        config: CompilerConfig, 
        packageReader: (name: string) => PackageConfig,
        logger: {
            info: (message: string) => void,
            error: (message: string) => void
        }
    ) {
        this.config = config;
        this.espidfComponents = new ESPIDFComponents(config.dirs.runtime);
        this.memory = new ShadowMemory(memoryLayout);
        this.packageReader = packageReader;
        this.logger = logger;
    }

    public async compile(): Promise<ExecutableBinary> {
        const {mainPackage, subPackages, mainEntryPoint, subModuleEntryPoints} = this._transpile();
        await this._compile(mainPackage, subPackages);
        await this._link(mainPackage, subPackages, mainEntryPoint);
        return this._generateExecutableBinary(mainPackage, mainEntryPoint, subModuleEntryPoints);
    }

    private _transpile() {
        let compileId = 0;
        const stdSrc = fs.readFileSync(STD_MODULE_PATH(this.config), 'utf-8');
        const nameTable = transpile(++compileId, stdSrc, undefined).names;
        const modules = new Map<string, GlobalVariableNameTable>(); // <path, nameTable>
        const subModuleEntryPoints: string[] = [];

        const mp = this.packageReader('main');
        const mainPackage: PackageConfig = mp;
        const visitedPackages: PackageConfig[] = [];
        let currentPackage: PackageConfig = mainPackage;

        const importer = (fname: string) => {
            let relativePath: string;
            if (!fname.startsWith('.')) {
                let pkg = visitedPackages.find(p => p.name === fname)
                if (!pkg) { // not visited package
                    currentPackage = this.packageReader(fname);
                    visitedPackages.push(currentPackage);
                }
                relativePath = './index';
            } else {
                relativePath = fname;
            }

            const modulePath = MODULE_PATH(currentPackage, relativePath);
            const mod = modules.get(modulePath);
            if (mod)
                return mod;
            else {
                if (!fs.existsSync(modulePath)) {
                    throw new Error(`Cannot find ${modulePath}`);
                }
                // Transpile
                const src = fs.readFileSync(modulePath).toString();
                const id = ++compileId;
                const result = transpile(id, src, nameTable, importer, id);
                modules.set(modulePath, result.names);
                subModuleEntryPoints.push(result.main);
                // Save C code
                const cSaveDir = path.join(currentPackage.dirs.dist, path.parse(relativePath).root);
                fs.mkdirSync(cSaveDir, {recursive: true});
                fs.writeFileSync(MODULE_C_PATH(cSaveDir, path.parse(relativePath).name), C_PROLOG(this.config) + result.code);
                return result.names;
            }
        }
    
        try {
            this.logger.info('Transpiling...');
            fs.mkdirSync(mainPackage.dirs.dist, {recursive: true});
            const mainSrc = fs.readFileSync(MODULE_PATH(mainPackage, './index'), 'utf-8');
            const result = transpile(++compileId, mainSrc, nameTable, importer);
            fs.writeFileSync(MODULE_C_PATH(mainPackage.dirs.dist, 'index'), C_PROLOG(this.config) + result.code);
            return {mainPackage, subPackages: visitedPackages, mainEntryPoint: result.main, subModuleEntryPoints};
        } catch (error) {
            this.logger.error(`Failed to transpile: ${error}`);
            throw error;
        }  
    }

    private async _compile(mainPackage: PackageConfig, subPackages: PackageConfig[]) {
        this.logger.info('Compiling with GCC...');

        // Compile each package with make.
        const commonIncludeDirs = this.espidfComponents.commonIncludeDirs;
        for (const pkg of [mainPackage, ...subPackages]) {
            try {
                this.logger.info(`Compiling ${pkg.name}...`);
                const makefile = generateMakefile(
                    this.config.dirs.compilerToolchain,
                    pkg,
                    [...this.espidfComponents.getIncludeDirs(pkg.dependencies), ...commonIncludeDirs],
                    ARCHIVE_PATH(pkg)
                );
                fs.writeFileSync(MAKEFILE_PATH(pkg), makefile);
                await executeCommand('make', pkg.dirs.dist, false);
            } catch (error) {
                this.logger.error(`Failed to compile ${pkg.name}: ${error}`);
                throw error;
            }
        }
    }

    private async _link(mainPackage: PackageConfig, subPackages: PackageConfig[], mainEntryPoint: string) {
        this.logger.info(`Linking...`);
        try {
            const cwd = process.cwd();
            const elfReader = new ElfReader(RUNTIME_ELF_PATH(this.config));
            const symbolsInRuntime = elfReader.readAllSymbols().map(symbol => ({name: symbol.name, address: symbol.address}));
            const linkerscript = generateLinkerScript(
                [path.relative(cwd, ARCHIVE_PATH(mainPackage)), ...subPackages.map(pkg=>path.relative(cwd, ARCHIVE_PATH(pkg)))],
                this._getArchivesWithEspComponents(mainPackage, subPackages).map(ar => path.relative(cwd, ar)),
                this.memory,
                symbolsInRuntime,
                mainEntryPoint
            );
            fs.writeFileSync(LINKER_SCRIPT(mainPackage), linkerscript);
            await executeCommand(`${LD_PATH(this.config)} -o ${LINKED_ELF_PATH(mainPackage)} -T ${LINKER_SCRIPT(mainPackage)} --gc-sections`, cwd);
        } catch (error) {
            this.logger.error(`Failed to link: ${error}`);
            throw error;
        }
    }

    private _getArchivesWithEspComponents(mainPackage: PackageConfig, subPackages: PackageConfig[]): string[] {
        const espArchivesFromMain = this.espidfComponents.getArchiveFilePaths(mainPackage.espIdfComponents);
        const resultArchives = [ARCHIVE_PATH(mainPackage), ...espArchivesFromMain];
        const visitedEspArchives = new Set(espArchivesFromMain);
        const reversedPackages = subPackages.reverse();
        const addEspArchive = (espArchive: string) => {
            if (!visitedEspArchives.has(espArchive)) {
                resultArchives.push(espArchive);
                visitedEspArchives.add(espArchive);
            }
        }
        for (const pkg of reversedPackages) {
            resultArchives.push(ARCHIVE_PATH(pkg));
            const espArchivesFromPkg = this.espidfComponents.getArchiveFilePaths(pkg.espIdfComponents);
            espArchivesFromPkg.forEach(ar => addEspArchive(ar));
        }
        this.espidfComponents.commonArchiveFilePaths.forEach(ar => addEspArchive(ar));
        return resultArchives;
    }

    private _generateExecutableBinary(
        mainPackage: PackageConfig, mainEntryPointName: string, subEntryPointNames: string[]
    ): ExecutableBinary {
        const linkedElf = new ElfReader(LINKED_ELF_PATH(mainPackage));
        const definedSymbols = linkedElf.readDefinedSymbols();
        const iramSection = linkedElf.readSectionByName(this.memory.iram.name);
        const dramSection = linkedElf.readSectionByName(this.memory.dram.name);
        const iflashSection = linkedElf.readSectionByName(this.memory.iflash.name);
        const dflashSection = linkedElf.readSectionByName(this.memory.dflash.name);
        const entryPoints: {id: number, address: number}[] = [];
        const setEntryPoint = (id: number, name: string) => {
            const symbol = definedSymbols.find(s => s.name === name);
            if (symbol) { entryPoints.push({id, address: symbol.address}); }
            else { throw new Error(`Cannot find ${name} in executable elf.`) }
        }
        subEntryPointNames.forEach(epn => setEntryPoint(-1, epn));
        setEntryPoint(0, mainEntryPointName);
        return {
            iram: iramSection ? {address: iramSection.address, data: iramSection.value} : undefined,
            dram: dramSection ? {address: dramSection.address, data: dramSection.value} : undefined,
            iflash: iflashSection ? {address: iflashSection.address, data: iflashSection.value} : undefined,
            dflash: dflashSection ? {address: dflashSection.address, data: dflashSection.value} : undefined,
            entryPoints
        }
    }
}


class ESPIDFComponents {
    public readonly commonIncludeDirs: string[];
    public readonly commonArchiveFilePaths: string[];

    private readonly COMPONENTS_PATH_PREFIX = /^.*microcontroller/;
    private readonly COMMON_COMPONENTS = ["cxx", "newlib", "freertos", "esp_hw_support", "heap", "log", "soc", "hal", "esp_rom", "esp_common", "esp_system"];
    private readonly RUNTIME_DIR: string; 
    private readonly SDK_CONFIG_DIR: string;

    private dependenciesInfo: {
        [key: string]: {
            file: string,
            dir: string,
            reqs: string[], 
            priv_reqs: string[],
            include_dirs: string[]
    }};

    constructor(runtimeDir: string) {
        this.RUNTIME_DIR = runtimeDir;
        this.SDK_CONFIG_DIR = path.join(runtimeDir, 'ports/esp32/build/config');
        const dependenciesFile = path.join(runtimeDir, 'ports/esp32/build/project_description.json');
        this.dependenciesInfo = JSON.parse(fs.readFileSync(dependenciesFile).toString()).build_component_info;
        this.commonIncludeDirs = this.getIncludeDirs(this.COMMON_COMPONENTS);
        this.commonArchiveFilePaths = this.getArchiveFilePaths(this.COMMON_COMPONENTS);
    }

    public getIncludeDirs(rootComponentNames: string[]) {
        const components = this._getComponents(rootComponentNames);
        const includeDirs:string[] = [];
        for (const component of components) {
            for (const dir of component.include_dirs) {
                includeDirs.push(path.join(component.dir, dir));
            }
        }
        includeDirs.push(this.SDK_CONFIG_DIR);
        return includeDirs;
    }

    public getArchiveFilePaths(rootComponentNames: string[]) {
        return this._getComponents(rootComponentNames).map(c => c.file);
    }

    private _getComponents(rootComponentNames: string[]) {
        let tmp = [...rootComponentNames];
        let visited = new Set<string>();
        const components = [];
        while(tmp.length > 0) {
        let curr = tmp.shift() as string;
        visited.add(curr)
        tmp = tmp.concat(
            this.dependenciesInfo[curr].priv_reqs.filter((r:string) => !visited.has(r)),
            this.dependenciesInfo[curr].reqs.filter((r:string) => !visited.has(r))
        );
        if (this.dependenciesInfo[curr].file !== undefined && this.dependenciesInfo[curr].file !== '')
            components.push(this.dependenciesInfo[curr]);
        }
        return components;
    }

    private _convertRuntimeDirPath(absolutePath: string) {
        return absolutePath.replace(this.COMPONENTS_PATH_PREFIX, this.RUNTIME_DIR);
    }
}


function executeCommand(command: string, cwd?: string, showStdout:boolean=true): Promise<void> {
  return new Promise((resolve, reject) => {
    const executeProcess = spawn(command, {shell: true, cwd});

    executeProcess.stdout.on('data', (data) => {
        if (showStdout) {
            process.stdout.write(data.toString());
        }
    });

    executeProcess.stderr.on('data', (data) => {
        process.stderr.write(data.toString());
    });

    executeProcess.on('error', (err) => {
        reject(new Error(`Failed to execute ${command}: ${err.message}`));
    });

    executeProcess.on('close', (code) => {
        if (code === 0) {
            resolve();
        } else {
            reject(new Error(`Failed to execute ${command}. Code: ${code}`));
        }
    });
  });
}



