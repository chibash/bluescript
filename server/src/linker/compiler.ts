import {MemoryBlock, MemoryRegion, ReusableMemoryRegion} from "./memory-region";
import {ElfReader, SECTION_TYPE, Symbol} from "./elf-reader";
import * as fs from "fs";
import {execSync} from "child_process";
import {
  LinerScriptMemoryAttribute,
  LinkerScript3,
  LinkerScriptMemoryRegion,
  LinkerScriptSection
} from "./linker-script3";


export type MemoryAddresses = {
  iram:{address:number, size:number},
  dram:{address:number, size:number},
  flash:{address:number, size:number},
}

type MemoryUpdate = {
  blocks: {address: number, data: string, isFlash: boolean}[],
  entryPoints: number[]
}

const EXTERNAL_SYMBOL_SECTION = '.external_symbols'

export class ShadowMemory {
  public readonly iram: ReusableMemoryRegion;
  public readonly dram: MemoryRegion;
  public readonly flash: MemoryRegion;
  public readonly symbols:Map<string, Symbol> = new Map<string, Symbol>();
  public readonly componentsInfo = new ESPIDFComponentsInfo();
  private updates: MemoryUpdate = {blocks: [], entryPoints: []};

  constructor(bsRuntimePath: string, addresses: MemoryAddresses) {
    const bsRuntime = new ElfReader(bsRuntimePath);
    bsRuntime.readAllSymbols().forEach(symbol => {this.symbols.set(symbol.name, symbol)});
    this.iram = new ReusableMemoryRegion('IRAM', addresses.iram.address, addresses.iram.size);
    this.dram = new MemoryRegion('DRAM', addresses.dram.address, addresses.dram.size);
    this.flash = new MemoryRegion('Flash', addresses.flash.address, addresses.flash.size);
  }

  setUpdate(blocks: {address: number, data: string, isFlash: boolean}[], entryPoint: number) {
    this.updates.blocks = this.updates.blocks.concat(blocks);
    this.updates.entryPoints.push(entryPoint)
  }

  getUpdates() {
    const updates = this.updates;
    this.updates = {blocks: [], entryPoints: []};
    return updates;
  }
}


export class Compiler {
  private readonly C_FILE = './temp-files/code.c';
  private readonly OBJ_FILE = './temp-files/code.o';
  private readonly LINKER_SCRIPT = './temp-files/linkerscript.ld';
  private readonly LINKED_ELF = './temp-files/code';
  constructor(private compilerPath = '') {}

  compile(shadowMemory: ShadowMemory, src: string, entryPointName: string) {
    const objFile = this._compile(src);
    const linkedElf = this.link(shadowMemory, objFile, entryPointName);
    this.load(shadowMemory, linkedElf, entryPointName)
  }

  private _compile(src: string) {
    fs.writeFileSync(this.C_FILE, src);
    execSync(`${this.compilerPath}xtensa-esp32-elf-gcc -c -O2 ${this.C_FILE} -o ${this.OBJ_FILE} -w -fno-common -ffunction-sections -mtext-section-literals -mlongcalls`);
    return new ElfReader(this.OBJ_FILE);
  }

  private link(shadowMemory: ShadowMemory, objFile: ElfReader, entryPointName: string) {
    this.freeIram(shadowMemory, objFile);
    const memoryBlocks = this.allocate(shadowMemory, objFile);
    const externalSymbols = this.readExternalSymbols(shadowMemory, objFile);
    const linkerScript = this.generateLinkerScript(memoryBlocks.iram, memoryBlocks.dram, externalSymbols, objFile, entryPointName);
    const linkedElf = this._link(linkerScript);
    this.freeEntryPoint(shadowMemory, entryPointName);
    return linkedElf;
  }

  private _link(linkerScript: LinkerScript3) {
    linkerScript.save(this.LINKER_SCRIPT)
    execSync(`${this.compilerPath}xtensa-esp32-elf-ld -o ${this.LINKED_ELF} -T ${this.LINKER_SCRIPT}`);
    return new ElfReader(this.LINKED_ELF);
  }

  private load(shadowMemory: ShadowMemory, linkedElf: ElfReader, entryPointName: string) {
    const blocks = linkedElf
      .readAllSections()
      .filter(section => section.name !== EXTERNAL_SYMBOL_SECTION)
      .map(section => ({address: section.address, data: section.value.toString("hex"), isFlash: false}));
    linkedElf.readDefinedSymbols().forEach(symbol => {
      shadowMemory.symbols.set(symbol.name, symbol);
    });
    const entryPoint = shadowMemory.symbols.get(entryPointName)?.address;
    if (entryPoint === undefined)
      throw new Error(`Cannot find entry point: ${entryPointName}`);
    shadowMemory.setUpdate(blocks, entryPoint);
  }

  private readExternalSymbols(shadowMemory: ShadowMemory, objFile: ElfReader) {
    const externalSymbols: Symbol[] = [];
    objFile.readExternalSymbols().forEach(symbol => {
      const definedSymbol = shadowMemory.symbols.get(symbol.name);
      if (definedSymbol !== undefined)
        externalSymbols.push(definedSymbol);
    });
    return externalSymbols;
  }

  private generateLinkerScript(
    iramBlocks: MemoryBlock[], dramBlocks: MemoryBlock[],
    externalSymbols: Symbol[], objFile: ElfReader, entryPointName: string) {
    const memories: LinkerScriptMemoryRegion[] = [];
    const sections: LinkerScriptSection[] = [];
    const externalSymbolMemory = new LinkerScriptMemoryRegion(
      'EXTERNAL_SYMBOLS',
      [new LinerScriptMemoryAttribute('executable')],
      0, 0
    );
    const externalSymbolSection = new LinkerScriptSection(EXTERNAL_SYMBOL_SECTION, externalSymbolMemory);
    externalSymbols.forEach(symbol => {
      externalSymbolSection.symbol(symbol.name, symbol.address);
    });
    memories.push(externalSymbolMemory);
    sections.push(externalSymbolSection);

    iramBlocks.forEach((block , index) => {
      const memoryRegion = new LinkerScriptMemoryRegion(
        `IRAM${index}`,
        [new LinerScriptMemoryAttribute('executable')],
        block.address, block.size);
      const section = new LinkerScriptSection(`.iram${index}`, memoryRegion)
        .section(objFile.filePath, [block.sectionName ?? '']);
      memories.push(memoryRegion);
      sections.push(section);
    });
    dramBlocks.forEach((block , index) => {
      const memoryRegion = new LinkerScriptMemoryRegion(
        `DRAM${index}`,
        [new LinerScriptMemoryAttribute('read/write'), new LinerScriptMemoryAttribute('readonly')],
        block.address, block.size);
      const section = new LinkerScriptSection(`.dram${index}`, memoryRegion)
        .section(objFile.filePath, [block.sectionName ?? '']);
      memories.push(memoryRegion);
      sections.push(section);
    });

    return new LinkerScript3()
      .input([objFile.filePath])
      .entry(entryPointName)
      .memory(memories)
      .sections(sections)
  }

  private allocate(shadowMemory: ShadowMemory, objFile: ElfReader) {
    const executableSections = objFile.readSections(SECTION_TYPE.EXECUTABLE).filter(section => section.size !== 0);
    const dataSections = objFile.readSections(SECTION_TYPE.WRITABLE)
      .concat(objFile.readSections(SECTION_TYPE.READONLY))
      .filter(section => section.size !== 0);
    const iramMemoryBlock = executableSections.map(section => shadowMemory.iram.allocate(section));
    const dramMemoryBlock = dataSections.map(section => shadowMemory.dram.allocate(section));
    return {iram: iramMemoryBlock, dram: dramMemoryBlock};
  }

  private freeEntryPoint(shadowMemory: ShadowMemory, entryPointName: string) {
    shadowMemory.iram.free(this.funcNameToSectionName(entryPointName));
  }

  private freeIram(shadowMemory: ShadowMemory, objFile: ElfReader) {
    const funcSymbols = objFile.readFunctions();
    funcSymbols.forEach(symbol => {
      if (shadowMemory.symbols.has(symbol.name))
        shadowMemory.iram.free(this.funcNameToSectionName(symbol.name))
    })
  }

  private funcNameToSectionName(funcName: string) {
    return `.text.${funcName}`
  }
}

export class ModuleCompiler {
  private readonly LINKER_SCRIPT = './temp-files/module-linkerscript.ld';
  private readonly LINKED_ELF = './temp-files/module-code';
  private readonly EXECUTABLE_SECTIONS = ['.literal', '.text', '.literal.*', '.text.*', '.iram*'];
  private readonly DATA_SECTIONS = ['.data', '.data.*', '.rodata', '.rodata.*', '.bss', '.bss.*', '.dram*'];
  private readonly DRAM_SECTION_NAME = '.dram';
  private readonly FLASH_SECTION_NAME = '.flash';


  constructor(private compilerPath: string = '') {}

  compile(shadowMemory: ShadowMemory, moduleName: string, entryPointName: string) {
    const executableElf = this.link(shadowMemory, moduleName, entryPointName);
    this.load(shadowMemory, executableElf, entryPointName)
  }

  private link(shadowMemory: ShadowMemory, moduleName: string, entryPointName: string): ElfReader {
    const linkerScript = this.generateLinkerScript(shadowMemory, moduleName, entryPointName);
    const linkedElf = this._link(linkerScript);
    this.allocate(shadowMemory, linkedElf);
    return linkedElf;
  }

  private _link(linkerScript: LinkerScript3) {
    linkerScript.save(this.LINKER_SCRIPT);
    execSync(`${this.compilerPath}xtensa-esp32-elf-ld -o ${this.LINKED_ELF} -T ${this.LINKER_SCRIPT} --gc-sections`);
    return new ElfReader(this.LINKED_ELF);
  }

  private load(shadowMemory: ShadowMemory, linkedElf: ElfReader, entryPointName: string) {
    const dramSection = linkedElf.readSectionByName(this.DRAM_SECTION_NAME);
    const flashSection = linkedElf.readSectionByName(this.FLASH_SECTION_NAME);
    const blocks: {address: number, data: string, isFlash: boolean}[] = []
    if (dramSection !== undefined)
      blocks.push({address: dramSection.address, data: dramSection.value.toString("hex"), isFlash: false});
    if (flashSection !== undefined)
      blocks.push({address: flashSection.address, data: flashSection.value.toString("hex"), isFlash: true});
    linkedElf.readDefinedSymbols().forEach(symbol => {
      shadowMemory.symbols.set(symbol.name, symbol);
    });
    const entryPoint = shadowMemory.symbols.get(entryPointName)?.address;
    if (entryPoint === undefined)
      throw new Error(`Cannot find entry point: ${entryPointName}`);
    shadowMemory.setUpdate(blocks, entryPoint);
  }

  private allocate(shadowMemory: ShadowMemory, linkedElf: ElfReader) {
    const dramSection = linkedElf.readSectionByName(this.DRAM_SECTION_NAME);
    const flashSection = linkedElf.readSectionByName(this.FLASH_SECTION_NAME);
    if (dramSection !== undefined)
      shadowMemory.dram.allocate(dramSection);
    if (flashSection !== undefined)
      shadowMemory.flash.allocate(flashSection);
  }

  private generateLinkerScript(shadowMemory: ShadowMemory, moduleName: string, entryPointName: string) {
    const moduleObjFile = shadowMemory.componentsInfo.getComponentPath(moduleName)
    const dramMemory = new LinkerScriptMemoryRegion(
      'DRAM',
      [new LinerScriptMemoryAttribute('read/write'), new LinerScriptMemoryAttribute('readonly')],
      shadowMemory.dram.getNextAddress(),
      1000000);
    const flashMemory = new LinkerScriptMemoryRegion(
      'FLASH',
      [new LinerScriptMemoryAttribute('executable')],
      shadowMemory.flash.getNextAddress(),
      1000000);
    const externalSymbolMemory = new LinkerScriptMemoryRegion(
      'EXTERNAL_SYMBOLS',
      [new LinerScriptMemoryAttribute('executable')],
      0, 0)
    const memories = [dramMemory, flashMemory, externalSymbolMemory];
    const externalSymbolSection = new LinkerScriptSection(EXTERNAL_SYMBOL_SECTION, externalSymbolMemory);
    shadowMemory.symbols.forEach(symbol => {
      externalSymbolSection.symbol(symbol.name, symbol.address);
    });
    const sections = [
      new LinkerScriptSection(this.DRAM_SECTION_NAME, dramMemory)
        .section(moduleObjFile, this.DATA_SECTIONS, true)
        .section('*', this.DATA_SECTIONS),
      new LinkerScriptSection(this.FLASH_SECTION_NAME, flashMemory)
        .section(moduleObjFile, this.EXECUTABLE_SECTIONS, true)
        .section('*', this.EXECUTABLE_SECTIONS),
      externalSymbolSection
    ];
    return new LinkerScript3()
      .input(shadowMemory.componentsInfo.getComponentsPath(moduleName))
      .memory(memories)
      .sections(sections)
      .entry(entryPointName)
  }
}


class ESPIDFComponentsInfo {
  static DEPENDENCIES_FILE_PATH = '../microcontroller/ports/esp32/build/project_description.json'
  static COMPONENTS_PATH_PREFIX = /^.*microcontroller\/ports\/esp32\/build/
  static RELATIVE_PATH_COMMON = '../microcontroller/ports/esp32/build'

  private dependenciesInfo: {[key: string]: {file: string, reqs: string[], priv_reqs: string[]}}

  constructor() {
    this.dependenciesInfo = JSON.parse(fs.readFileSync(ESPIDFComponentsInfo.DEPENDENCIES_FILE_PATH).toString()).build_component_info
  }

  public getComponentsPath(rootComponent: string): string[] {
    let tmp = [rootComponent]
    let visited = new Set<string>()
    const componentPaths:string[] = []
    while(tmp.length > 0) {
      let curr = tmp.shift() as string
      visited.add(curr)
      tmp = tmp.concat(
        this.dependenciesInfo[curr].priv_reqs.filter((r:string) => !visited.has(r)),
        this.dependenciesInfo[curr].reqs.filter((r:string) => !visited.has(r))
      )
      if (this.dependenciesInfo[curr].file !== undefined && this.dependenciesInfo[curr].file !== '')
        componentPaths.push(this.convertAbsoluteToRelative(this.dependenciesInfo[curr].file))
    }
    return componentPaths
  }

  public getComponentPath(componentName: string) {
    return this.convertAbsoluteToRelative(this.dependenciesInfo[componentName].file)
  }

  private convertAbsoluteToRelative(absolutePath: string) {
    return absolutePath.replace(ESPIDFComponentsInfo.COMPONENTS_PATH_PREFIX, ESPIDFComponentsInfo.RELATIVE_PATH_COMMON)
  }
}

