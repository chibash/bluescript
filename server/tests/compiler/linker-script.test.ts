import { describe, expect, it } from '@jest/globals';
import { ShadowMemory } from '../../src/compiler/compiler';
import generateLinkerScript from '../../src/compiler/linker-script';

describe('generateLinkerScript', () => {

  const mockShadowMemory: ShadowMemory = {
    iram: { name: '.iram', address: 0x40080000, size: 10000, used:0 },
    dram: { name: '.dram', address: 0x3FFB0000, size: 10000, used:0 },
    iflash: { name: '.iflash', address: 0x400D0000, size: 1000, used:0 },
    dflash: { name: '.dflash', address: 0x3F400000, size: 1000, used:0 },
  };

  it('should generate a correct linker script with all parameters provided', () => {
    const inputFiles = ['liba.a', 'libc.a', 'main.o', 'utils.o'];
    const definedSymbols = [
      { name: 'defined_func1', address: 0x12345678 },
      { name: 'defined_val1', address: 0x87654321 },
    ];
    const entryPointName = 'bluescript_main0_';
    const subModuleEntryPoints = ['bluescript_main1_0', 'bluescript_main2_0'];

    const script = generateLinkerScript(
      inputFiles,
      mockShadowMemory,
      definedSymbols,
      entryPointName,
      subModuleEntryPoints
    );

    expect(script).toMatchSnapshot();
  });
});