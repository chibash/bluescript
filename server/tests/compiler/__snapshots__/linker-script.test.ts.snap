// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateLinkerScript should generate a correct linker script with all parameters provided 1`] = `
"GROUP(liba.a libc.a main.o utils.o)
ENTRY(bluescript_main0_)


MEMORY {
	IRAM   (XA)   : ORIGIN = 0x40080000,  LENGTH = 1000000
	DRAM   (WA)   : ORIGIN = 0x3ffb0000,  LENGTH = 1000000
	IFlash   (X)   : ORIGIN = 0x400d0000,  LENGTH = 1000000
	DFlash   (R)   : ORIGIN = 0x3f400000,  LENGTH = 1000000
	EXTERNAL   (X)   : ORIGIN = 0x0,  LENGTH = 0
}


SECTIONS {
	.iram : {
		*(.iram*)
		KEEP(main.o(.iram*))
		KEEP(utils.o(.iram*))
		. = ALIGN(4);
	} > IRAM

	.dram : {
		*(.data .data.* .bss .bss.* .dram*)
		KEEP(main.o(.data .data.* .bss .bss.* .dram*))
		KEEP(utils.o(.data .data.* .bss .bss.* .dram*))
	} > DRAM

	.iflash : {
		*(.literal .text .literal.* .text.*)
		KEEP(main.o(.literal .text .literal.* .text.*))
		KEEP(utils.o(.literal .text .literal.* .text.*))
		. = ALIGN(4);
	} > IFlash

	.dflash : {
		*(.rodata .rodata.*)
		KEEP(main.o(.rodata .rodata.*))
		KEEP(utils.o(.rodata .rodata.*))
	} > DFlash

	.external : {
		defined_func1 = 0x12345678;
		defined_val1 = 0x87654321;
	} > EXTERNAL
}
"
`;
